language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/toml/common/completion/TomlCompletionUtil.java

```diff
@@ -0,0 +1,233 @@
+            fields = topLevelNode.getRight().fields();
+            documentNode = (DocumentNode) topLevelNode.getRight().parent();
+        }
+
+        Map<String, CompletionItem> completions =
+                new HashMap<>(findCompletionItemsFromQualifiedKey(snippets, tableKey));
+
+        /*   Removes the following.
+            eg: [build-options]
+                cloud="docker"
+            The completion item that corresponds to the above key-value pair
+            will be removed as it is already defined.
+         */
+        for (KeyValueNode field : fields) {
+            String key = TomlSyntaxTreeUtil.toQualifiedName(field.identifier().value());
+            completions.remove(key);
+        }
+
+        /*
+            Add all possible top-level (Table & TableArray) keys to the completion-item list.
+            And remove already existing keys from the completion item list.
+            eg: [build-options],[package]
+         */
+        addTopLevelNodeCompletions(snippets.keySet(), completions);
+        return removeExistingTableKeys(completions, documentNode);
+    }
+
+    /**
+     * Returns the set of  completion items for the given set of toml nodes.
+     *
+     * @param topLevelNodes The set of top-level toml nodes.
+     * @param completions   The set of completion items.
+     * @return
+     */
+    public static Map<String, CompletionItem> addTopLevelNodeCompletions(Set<TomlNode> topLevelNodes,
+                                                                            Map<String, CompletionItem> completions) {
+        for (TomlNode topLevelNode : topLevelNodes) {
+            String key = topLevelNode.getKey();
+            if (completions.containsKey(key)) {
+                continue;
+            }
+            CompletionItem item;
+            if (topLevelNode.type() == TomlNodeType.TABLE || topLevelNode.type() == TomlNodeType.TABLE_ARRAY) {
+                item = createTopLevelCompletionItem(topLevelNode);
+            } else {
+                continue;
+            }
+            completions.put(key, item);
+        }
+        return completions;
+    }
+
+    /**
+     * Removes already declared Table completion items from the completion item list.
+     *
+     * @param completions  The set completion items
+     * @param documentNode Toml document node.
+     * @return {@link Map<String,CompletionItem>} The optimized completion item list.
+     */
+    public static Map<String, CompletionItem> removeExistingTableKeys(Map<String, CompletionItem> completions,
+                                                                      DocumentNode documentNode) {
+        List<String> existingKeys = completions.entrySet().stream()
+                .filter(entry -> isTableDeclared(documentNode, entry.getKey()))
+                .map(Map.Entry::getKey).collect(Collectors.toList());
+        for (String key : existingKeys) {
+            completions.remove(key);
+        }
```

Comment:
May be we can try to simplify this and handle this in a single run (Not a major revamp for now).

Reply:
Done